backend/mem/location.go

```diff
@@ -0,0 +1,202 @@
+}
+
+/*
+ListByPrefix tags a prefix onto the current path and in a slice,
+ returns all file base names whose full paths contain that substring
+Returns empty slice if nothing found
+*/
+
+func (l *Location) ListByPrefix(prefix string) ([]string, error) {
+
+	list := make([]string, 0)
+	str := path.Join(l.Path(), prefix)
+	mapRef := &l.fileSystem.fsMap
+	if _, ok := (*mapRef)[l.volume]; ok {
+		paths := (*mapRef)[l.volume].getKeys()
+		for i := range paths {
+			if strings.Contains(paths[i], str) {
+				if path.Ext(paths[i]) != "" && strings.Contains(str, utils.EnsureTrailingSlash(path.Dir(paths[i]))) {
+					list = append(list, path.Base(paths[i]))
+					sort.Strings(list)
+				}
+			}
+		}
+	}
+
+	return list, nil
+}
+
+/*
+ListByRegex takes a regular expression and returns
+a slice of strings containing the base names of files
+found that matched the regular expression.  Returns an
+empty slice upon nothing found
+*/
+
+func (l *Location) ListByRegex(regex *regexp.Regexp) ([]string, error) {
+
+	list := make([]string, 0)
+	str := l.Path()
+	mapRef := &l.fileSystem.fsMap
+	if _, ok := (*mapRef)[l.Volume()]; ok {
+		namesHere := (*mapRef)[l.Volume()].fileNamesHere(str)
+		for i := range namesHere {
+			if regex.MatchString(namesHere[i]) {
+				list = append(list, namesHere[i])
+			}
+		}
+	}
+	return list, nil
+}
+
+//Volume returns the volume of the current FileSystem.
+func (l *Location) Volume() string {
+	return l.volume
+}
+
+//Path returns the full, absolute path of the location with leading and trailing slashes
+func (l *Location) Path() string {
+
+	//just to be sure that we return a trailing and leading slash
+	str := utils.EnsureTrailingSlash(path.Clean(l.name))
+	return utils.EnsureLeadingSlash(str)
+
+}
+
+//Exists always returns true on locations
+func (l *Location) Exists() (bool, error) {
+
+	l.exists = true
+	return true, nil
+}
+
+/*
+NewLocation creates a new location at the
+given relative path, which is tagged onto the current locations absolute path
+*/
+func (l *Location) NewLocation(relLocPath string) (vfs.Location, error) {
+	err := utils.ValidateRelativeLocationPath(relLocPath)
+	if err != nil {
+		return nil, err
+	}
+	str := path.Join(l.Path(), relLocPath)
+	str = utils.EnsureTrailingSlash(path.Clean(str))
+	return &Location{
+		fileSystem: l.fileSystem,
+		name:       str,
+		exists:     false,
+		volume:     l.Volume(),
+	}, nil
+
+}
+
+//ChangeDir simply changes the directory of the location
+func (l *Location) ChangeDir(relLocPath string) error {
+	err := utils.ValidateRelativeLocationPath(relLocPath)
+	if err != nil {
+		return err
+	}
+	l.name = path.Join(l.name, relLocPath)
+	return nil
+
+}
+
+//FileSystem returns the type of file system location exists on, if it exists at all
+func (l *Location) FileSystem() vfs.FileSystem {
+
+	return l.fileSystem
+
+}
+
+//NewFile creates a vfs.File given its relative path and tags it onto "l's" path
+func (l *Location) NewFile(relFilePath string) (vfs.File, error) {
+
```

Comment:
Same here. This function should check the file system to see if a file already exists at the location with the given name. Return early with that file's reference if it exists.

```
obj, ok := l.fileSystem.fsMap[l.volume][path.Join(l.Path(), relFilePath)]
if ok && obj.isFile {
		return obj.i.(*File), nil
}
```